slep000, slep workflow (#30)
scikit-learn/enhancement_proposals: slep000/proposal.rst

.. _slep_000:

==============================

> SLEP000: SLEP and its wofkflow

Reviewer: ```suggestion
SLEP000: SLEP and its workflow
```

==============================

:Author: Adrin Jalali
:Status: Draft
:Type: Process
:Created: 2020-02-13

Abstract
########

This SLEP specifies details related to SLEP submission, review, and acceptance
process.

Motivation
##########

Without a predefined workflow, the discussions around a SLEP can be long and

> frustrating for both the author and the reviewers. The lack of a known workflow

Reviewer 2: ```suggestion
frustrating for both the author(s) and the reviewers. The lack of a known workflow
```

also results in the SLEPs to take months (if not years) before it is accepted

> as ``Draft``. The purpose of this SLEP is to lubricate and ease the

Reviewer: ```suggestion
as ``Under Review``. The purpose of this SLEP is to lubricate and ease the
```

My understanding is that drafts are open PRs and Under Review means people are voting on it? That's also the definition that is used below

Author: no, there's no `under review` state for PRs. They are a draft as long as there are active changes done on them.

Reviewer: That's the current process at least but I understand you want to change that.

I'm confused by the formulation "accepted as draft", which suggest that there needs to be a review process for even *having* a draft. I would change that to 

> The lack of a known workflow
also results in the SLEPs to take months (if not years) for a SLEP to be accepted

Author: My issue is not that it takes a very long time to be accepted. My issue is that because the PRs get too long, it goes out of control. I changed it to "merged as Draft"

> as ``Draft``. The purpose of this SLEP is to lubricate and ease the

Reviewer: ```suggestion
as ``Draft``. The purpose of this SLEP is to facilitate and ease the
```

Author: I like the word _lubricate_, there's a lot of friction and this helps it.

Reviewer: Ok. Again, not trying to be a pain. My comment is motivated by https://www.google.com/search?q=lubricant&safe=off&client=ubuntu&hs=P7z&hl=en&sxsrf=ACYBGNRkj6suJB0x3HsGRfks0X-BaNIggw:1581611794246&source=lnms&tbm=isch&sa=X&ved=2ahUKEwjaoKyJ-87nAhUu2FkKHesuASoQ_AUoAnoECBgQBA&biw=1758&bih=991

process of working on a SLEP, and make it a more enjoyable and productive
experience.

What is a SLEP?
###############

> SLEP stands for Scikit-Learn Enhancement Proposal. A SLEP is a design document

Reviewer: ```suggestion
SLEP stands for Scikit-Learn Enhancement Proposal, inspired from Python PEPs or Numpy NEPs. A SLEP is a design document
```

providing information to the scikit-learn community, or describing a new
feature for scikit-learn or its processes or environment. The SLEP should

> provide a concise technical specification of the feature and a rationale for

Reviewer: ```suggestion
provide a concise technical specification of the proposed solution, and a rationale for
```

the feature.

We intend SLEPs to be the primary mechanisms for proposing major new features,
for collecting community input on an issue, and for documenting the design
decisions that have gone into scikit-learn. The SLEP author is responsible for
building consensus within the community and documenting dissenting opinions.

Because the SLEPs are maintained as text files in a versioned repository, their
revision history is the historical record of the feature proposal.

SLEP Audience
#############

The typical primary audience for SLEPs are the core developers of
``scikit-learn`` and technical committee, as well as contributors to the
project. However, these documents also serve the purpose of documenting the
changes and decisions to help users understand the changes and why they are
made.

The proposal should be submitted as a draft SLEP via a GitHub pull request to a
``slepXXX`` directory with the name ``proposal.rst`` where ``XXX`` is an
appropriately assigned three-digit number (e.g., ``slep000/proposal.rst``). The

> draft must use the SLEP — Template and Instructions file.

Reviewer: mention the https://github.com/scikit-learn/enhancement_proposals repo and the website

Once the PR for the SLEP is in place, a post should be made to the mailing list

> containing the sections up to “Backward compatibility”, with the purpose of

Reviewer: what does "in place" mean?

Author: exists?

Reviewer: OK, "created" then. "in place" suggests some sort of maturity in my mind

> limiting discussion there to usage and impact. Discussion on the pull request

Reviewer: Are you suggesting to also enable discussion on the mailing list? Wouldn't this scatter/fraction the discussion?

Author: Yes it would, but some of the discussions where mostly non-core-devs are involved and give feedback, doesn't have to be on the PR.

Reviewer 3: Given that this is also the NEP process that seems fine, I think?

will have a broader scope, also including details of implementation.

> At the earliest convenience, the PR should be merged (regardless of whether it

Reviewer: what is "earliest convenience"?

Should -> may (let's at least not make it a mandatory process please)

Author: This is the main reason I opened this SLEP. If there's one thing I would argue for, is this one.

Reviewer: Then what does "At the earliest convenience" mean?

Depending on the interpretation, that would mean you could just merge this PR right now. I doubt this is a productive workflow.

Also imagine a situation where I'm writing a SLEP PR, and you and I disagree on some point. I'm merging the PR, because I have to, according to this SLEP. Your points in our discussion are "lost" now. You have to make them again, in another PR. That's twice as much work for you.

Author: I would argue that right now is a really good point to merge it actually. And that applies to the feature names out, and the resampler one. They may need more work, but the whole point is that they should be merged now.

We can have issues and separate PRs discussing specific points, for instance, one for alternatives, one for abstract, etc.

Reviewer: Can you explain why you think it would ease the SLEP process to merge early and discuss in separate PRs?

In the end, comments will be made, and they will have to be addressed. I'm failing to see how having separate PRs will help. I understand the initial frustration was that on long PRs, it's hard to keep track of the changes and the discussions. But that doesn't seem to apply to all SLEPs, so I don't see why we should make this mandatory

Author: For the same reason that we break large PRs into several smaller ones in many cases. For a small PR, that point may come a bit later if the discussion doesn't get long. For a larger SLEP, that point comes earlier.

Reviewer: > For the same reason that we break large PRs into several smaller ones in many cases

This is where I disagree: we have merged PRs with *reduced scope*, but as far as I know, we do not merge PRs that aren't properly finished yet, with the hope that issues will be addressed later.

Typically, I consider https://github.com/scikit-learn/scikit-learn/pull/16112 to be one of these PRs with a reduced scope: it's  addressing one part of SLEP010. But I'm making sure it doesn't get merged until all the points that are raised (and that are in scope) are addressed.

Author: both python and numpy communities agree that when it comes to *EPs, they should get merged before they are _properly_ finished, and I agree with them.

Reviewer: I still don't believe this should be an obligation. But we don't have to agree

Reviewer 3: So the NEP says " Additional PRs may be made by the Author to update or expand the NEP, or by maintainers to set its status, discussion URL, etc.", meaning that the SLEP author is the only one who is supposed to change it, right?

I think it's fine to merge drafts, and it's up to the author to ensure the draft is mature enough to serve as a basis for discussion.

I'm not entirely sure how the process for NEPs and PEPs works in terms of the back and forth between the author and the community. I feel all the important points should be part of the SLEP so losing the detailed discussion on a PR might be ok?

Maybe we can ping the numpy people to tell us about their experience?
If I look at " Discussion on the pull request will have a broader scope, also including details of implementation." it sounds like the main discussion does happen on the pull request.

Reviewer 4: To comment on this point: I feel that my view sits a bit in the middle of that of @NicolasHug and of @adrinjalali :

* I really see the value of merging things that are consolidated but there still is open discussion
* We need them to be "consolidated" (we could punt on defining precisely what that means).

How about we say that they can be merged when the author considers them as ready for discussion and at least on core-dev is willing to champion this merge (that last statement is to avoid polluting the SLEPs repo with SLEPs that have no chances of making it).

is accepted during discussion). Additional PRs may be made by the champions to
update or expand the SLEP, or by maintainers to set its status, discussion URL,
etc.

> Standards Track SLEPs consist of two parts, a design document and a reference

Reviewer: ```suggestion
Standards Track SLEPs (see below) consist of two parts, a design document and a reference
```

Though I'm suggesting to remove the SLEP types section

implementation. It is generally recommended that at least a prototype
implementation be co-developed with the SLEP, as ideas that sound good in
principle sometimes turn out to be impractical when subjected to the test of
implementation. Often it makes sense for the prototype implementation to be
made available as PR to the scikit-learn repo (making sure to appropriately
mark the PR as a WIP).

SLEP Types

> ##########

Reviewer: I'm not sure this section is really useful and the doc is already pretty long

Author: This is needed to define the SLEPs, and is included both in PEP1 and NEP0, which is a shortened version of PEP1.

request on the main repo before submitting a SLEP. It is generally a good idea
for the author of the SLEP to gauge the viability and the interest of the
community before working on a SLEP, mostly to save author's time.

A SLEP must have one or more champions: people who write the SLEP following the
SLEP template, shepherd the discussions around it, and seek consensus in the
community.

Review and Resolution
---------------------

SLEPs are discussed on the mailing list or the PRs modifying the SLEP. The
possible paths of the status of SLEPs are as follows:

.. image:: pep-0001-process_flow.png
   :alt: SLEP process flow diagram

All SLEPs should be created with the ``Draft`` status.

Eventually, after discussion, there may be a consensus that the SLEP should be
accepted – see the next section for details. At this point the status becomes
``Accepted``.

Once a SLEP has been ``Accepted``, the reference implementation must be
completed. When the reference implementation is complete and incorporated into
the main source code repository, the status will be changed to ``Final``.

To allow gathering of additional design and interface feedback before

> committing to long term stability for a language feature or standard library

Reviewer 2: `language feature or standard library API` seems better for Python but not for scikit-learn. Or we should rephrase slightly.

A SLEP can also be assigned status ``Deferred``. The SLEP author or a core
developer can assign the SLEP this status when no progress is being made on the
SLEP.

A SLEP can also be ``Rejected``. Perhaps after all is said and done it was not
a good idea. It is still important to have a record of this fact. The
``Withdrawn`` status is similar; it means that the SLEP author themselves has
decided that the SLEP is actually a bad idea, or has accepted that a competing
proposal is a better alternative.

When a SLEP is ``Accepted``, ``Rejected``, or ``Withdrawn``, the SLEP should be
updated accordingly. In addition to updating the status field, at the very
least the ``Resolution`` header should be added with a link to the relevant
thread in the mailing list archives or where the discussion happened.

SLEPs can also be ``Superseded`` by a different SLEP, rendering the original
obsolete. The ``Replaced-By`` and ``Replaces`` headers should be added to the
original and new SLEPs respectively.

``Process`` SLEPs may also have a status of ``Active`` if they are never meant
to be completed, e.g. SLEP 1 (this SLEP).

How a SLEP becomes Accepted
---------------------------

> A SLEP is ``Accepted`` by the voting mechanism defined in the governance model.

Reviewer: link to it

We need a concrete way to tell whether consensus has been reached. When you
think a SLEP is ready to accept, create a PR changing the status of the SLEP to
``Accepted``, then send an email to the scikit-learn mailing list with a
subject like:

    Proposal to accept SLEP #<number>: <title>

In the body of your email, you should:

- link to the latest version of the SLEP, and a link to the PR accepting the
  SLEP.

- briefly describe any major points of contention and how they were resolved,

> - include a sentence like: “The vote will be closed in a month.”

Reviewer: ```suggestion
- include a sentence like: “The vote will be closed in a month i.e. on <the_date>.”
```

.. _slep_000:

==============================
SLEP000: SLEP and its workflow
==============================

:Author: Adrin Jalali
:Status: Draft
:Type: Process
:Created: 2020-02-13

Abstract
########

This SLEP specifies details related to SLEP submission, review, and acceptance
process.

Motivation
##########

Without a predefined workflow, the discussions around a SLEP can be long and
frustrating for both the author(s) and the reviewers. The lack of a known
workflow also results in the SLEPs to take months (if not years) before it is

> merged as ``Draft``. The purpose of this SLEP is to lubricate and ease the

Reviewer: But that does not currently apply.

We don't merge PRs as "draft" right now. We merge PRs that we think are ready to be voted on.

.. _slep_000:

==============================
SLEP000: SLEP and its workflow
==============================

:Author: Adrin Jalali
:Status: Draft
:Type: Process
:Created: 2020-02-13

Abstract
########

This SLEP specifies details related to SLEP submission, review, and acceptance
process.

Motivation
##########

Without a predefined workflow, the discussions around a SLEP can be long and

> frustrating for both the author(s) and the reviewers. The lack of a known

Reviewer 4: ```suggestion
consume a lot of energy for both the author(s) and the reviewers. The lack of a known
```
I like always avoiding too negative words

workflow also results in the SLEPs to take months (if not years) before it is
merged as ``Under Review``. The purpose of this SLEP is to lubricate and ease
the process of working on a SLEP, and make it a more enjoyable and productive

> experience. It also used the process used in PEPs and NEPs which means there

Reviewer 3: "uses"? I'm not sure I understand the sentence as it is written right now.

scikit-learn community, or describing a new feature for scikit-learn or its
processes or environment. The SLEP should provide a concise technical
specification of the proposed solution, and a rationale for the feature.

We intend SLEPs to be the primary mechanisms for proposing major new features,
for collecting community input on an issue, and for documenting the design
decisions that have gone into scikit-learn. The SLEP author is responsible for
building consensus within the community and documenting dissenting opinions.

Because the SLEPs are maintained as text files in a versioned repository, their
revision history is the historical record of the feature proposal.

SLEP Audience
#############

The typical primary audience for SLEPs are the core developers of
``scikit-learn`` and technical committee, as well as contributors to the
project. However, these documents also serve the purpose of documenting the
changes and decisions to help users understand the changes and why they are
made. The SLEPs are available under `Scikit-learn enhancement proposals
<https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/>`_.

> 

Reviewer 4: Should we have a new section here? I was personally surprised by the flow.

Author: right, moved it to the right section

request on the main repo before submitting a SLEP. It is generally a good idea
for the author of the SLEP to gauge the viability and the interest of the
community before working on a SLEP, mostly to save author's time.

A SLEP must have one or more champions: people who write the SLEP following the
SLEP template, shepherd the discussions around it, and seek consensus in the
community.

The proposal should be submitted as a draft SLEP via a GitHub pull request to a
``slepXXX`` directory with the name ``proposal.rst`` where ``XXX`` is an
appropriately assigned three-digit number (e.g., ``slep000/proposal.rst``). The
draft must use the `SLEP — Template and Instructions
<https://github.com/scikit-learn/enhancement_proposals/blob/master/slep_template.rst>`_
file.

Once the PR for the SLEP is created, a post should be made to the mailing list
containing the sections up to “Backward compatibility”, with the purpose of
limiting discussion there to usage and impact. Discussion on the pull request
will have a broader scope, also including details of implementation.

At the earliest convenience, the PR should be merged (regardless of whether it

> is accepted during discussion). Additional PRs may be made by the champions to

Reviewer 3: Maybe just add one sentence here like "It is expected the author and reviewers ensure the SLEP is complete and sound before merging."?

prototype implementation be co-developed with the SLEP, as ideas that sound
good in principle sometimes turn out to be impractical when subjected to the
test of implementation. Often it makes sense for the prototype implementation
to be made available as PR to the scikit-learn repo (making sure to
appropriately mark the PR as a WIP).

Review and Resolution
---------------------

SLEPs are discussed on the mailing list or the PRs modifying the SLEP. The
possible paths of the status of SLEPs are as follows:

.. image:: pep-0001-process_flow.png
   :alt: SLEP process flow diagram

All SLEPs should be created with the ``Draft`` status.

Eventually, after discussion, there may be a consensus that the SLEP should be
accepted – see the next section for details. At this point the status becomes
``Accepted``.

Once a SLEP has been ``Accepted``, the reference implementation must be
completed. When the reference implementation is complete and incorporated into
the main source code repository, the status will be changed to ``Final``.

To allow gathering of additional design and interface feedback before
committing to long term stability for a feature or API, a SLEP may also be

> marked as ``Provisional``. This is short for“Provisionally Accepted”, and

Reviewer 3: Is the formatting here correct?

subject like:

    Proposal to accept SLEP #<number>: <title>

In the body of your email, you should:

- link to the latest version of the SLEP, and a link to the PR accepting the
  SLEP.

- briefly describe any major points of contention and how they were resolved,

- include a sentence like: “The vote will be closed in a month i.e. on
  <the_date>.”

Generally the SLEP author will be the one to send this email, but anyone can do
it; the important thing is to make sure that everyone knows when a SLEP is on
the verge of acceptance, and give them a final chance to respond.

In general, the goal is to make sure that the community has consensus, not
provide a rigid policy for people to try to game. When in doubt, err on the
side of asking for more feedback and looking for opportunities to compromise.

If the final comment and voting period passes with the required majority, then
the SLEP can officially be marked ``Accepted``. The ``Resolution`` header
should link to the PR accepting the SLEP.

If the vote does not achieve a required majority, then the SLEP remains in
``Draft`` state, discussion continues as normal, and it can be proposed for

> acceptance again later once the objections are resolved.

Reviewer 4: Do we need a mechanism for rejecting clearly a SLEP?

The reason that I ask this is that I can be useful to avoid people loosing a lot of time on SLEP that is considered as not having chances of getting through?

Maybe the "rejection" could and should be done via a pull request to the SLEP that explains the reason for rejecting it?

Author: The SLEP in its current state says:

> The ``Withdrawn`` status is similar; it means that the SLEP author themselves has
> decided that the SLEP is actually a bad idea, or has accepted that a competing
> proposal is a better alternative.

I think it's important for the SLEP authors to make that decision if they wish, and otherwise leave it to the vote. rejected and withdrawn states are effectively the same. WDYT?

Author: In other words, if not by the author's decision (not necessarily exhaustion) and not the vote, how else would you like to be able to reject a SLEP?

Reviewer 3: I agree with @adrinjalali I think the mechanisms described here should be enough.

Reviewer 5: In the case where the original author does not feel like withdrawing a SLEP after a failed vote and the majority things that it should be withdrawn or rejected, what should we do?

Is it acceptable for someone else to do a PR to reject the slep and then use the usual decision making described in the governance doc to deal with the rejection PR?

If so maybe we should make it a bit more explicit here. Maybe just amending the next section to:

```
In unusual cases, with the request of the author, the scikit-learn technical
committee may be asked to decide whether a controversial SLEP is ``Accepted``, put back to ``Draft`` with additional recommendation to try again to reach consensus or definitely ``Rejected``. See the governance doc for more details.
```

Reviewer 5: In the case where the original author does not feel like withdrawing a SLEP after a failed vote and the majority things that it should be withdrawn or rejected, what should we do?

Is it acceptable for someone else to do a PR to reject the slep and then use the usual decision making described in the governance doc to deal with the rejection PR?

If so maybe we should make it a bit more explicit here. Maybe just amending the next section to:

```
In unusual cases, with the request of the author, the scikit-learn technical
committee may be asked to decide whether a controversial SLEP is ``Accepted``, put back to ``Draft`` with additional recommendation to try again to reach consensus or definitely ``Rejected``.
Please refer to the governance doc for more details.
```

Reviewer 3: hm so the ambiguity on how to reach the "rejected" state is copied from the NEP. The PEP doesn't have that problem because they basically delegate the decision to an individual that then makes the call and can reject, or the steering council can reject. With the NEP version that goes back to draft there's no way to get rejected. Anyway I think it's fine the way it is for now, we can always clarify later.

for the author of the SLEP to gauge the viability and the interest of the
community before working on a SLEP, mostly to save author's time.

A SLEP must have one or more champions: people who write the SLEP following the
SLEP template, shepherd the discussions around it, and seek consensus in the
community.

The proposal should be submitted as a draft SLEP via a GitHub pull request to a
``slepXXX`` directory with the name ``proposal.rst`` where ``XXX`` is an
appropriately assigned three-digit number (e.g., ``slep000/proposal.rst``). The
draft must use the `SLEP — Template and Instructions
<https://github.com/scikit-learn/enhancement_proposals/blob/master/slep_template.rst>`_
file.

Once the PR for the SLEP is created, a post should be made to the mailing list
containing the sections up to “Backward compatibility”, with the purpose of
limiting discussion there to usage and impact. Discussion on the pull request
will have a broader scope, also including details of implementation.

The SLEP is reviewed for structure, formatting, and other errors. Approval
criteria are:

- It sound and complete. The ideas must make technical sense.

> - The reviewers do not consider whether they seem likely to be accepted.

Reviewer: I don't understand this: this doesn't seem to be a criteria?

Author: I'm not sure if I understand the question. Maybe going through PEP1 would help?

- The title accurately describes the content.

Reviewers are generally quite lenient about this initial review, expecting that
problems will be corrected by the further reviewing process. **Note**: Approval
of the SLEP is no guarantee that there are no embarrassing mistakes! Ideally
they're avoided, but they can also be fixed later in separate PRs. Once
approved by at least one core developer, the SLEP can be merged. Additional PRs
may be made by the champions to update or expand the SLEP, or by maintainers to

> set its status, discussion URL, etc.

Reviewer: It's not clear to me whether this section refers to approval of the first PR, or actual approval of the SLEP. I assume it's the former, but the text often suggests the latter.

Author: This part is about approval of the PR.

Reviewer: Therefore I would kindly suggest to edit the parts were the text suggest that this is about SLEP approval:

- The SLEP is reviewed for structure, formatting, and other errors. Approval
criteria are...

- Approval of the SLEP...

- Once approved by at least one core developer, the SLEP...

containing the sections up to “Backward compatibility”, with the purpose of
limiting discussion there to usage and impact. Discussion on the pull request
will have a broader scope, also including details of implementation.

The first draft of the SLEP needs to be approved by at least one core developer
before being merged. Merging the draft does not mean it is accepted or is ready
for the vote. To this end, the SLEP draft is reviewed for structure,
formatting, and other errors. Approval criteria are:

- It sound and complete. The ideas must make technical sense.
- The reviewers do not consider whether they seem likely to be accepted.
- The title accurately describes the content.

Reviewers are generally quite lenient about this initial review, expecting that
problems will be corrected by the further reviewing process. **Note**: Approval
of the SLEP is no guarantee that there are no embarrassing mistakes! Ideally
they're avoided, but they can also be fixed later in separate PRs. Once
approved by at least one core developer, the SLEP can be merged. Additional PRs
may be made by the champions to update or expand the SLEP, or by maintainers to

> set its status, discussion URL, etc.

Reviewer: This section still mentions "SLEP approval" instead of "PR approval".

maintainers to set its status, discussion URL, etc.

Standards Track SLEPs (see bellow) consist of two parts, a design document and
a reference implementation. It is generally recommended that at least a
prototype implementation be co-developed with the SLEP, as ideas that sound
good in principle sometimes turn out to be impractical when subjected to the
test of implementation. Often it makes sense for the prototype implementation
to be made available as PR to the scikit-learn repo (making sure to
appropriately mark the PR as a WIP).

Review and Resolution
---------------------

SLEPs are discussed on the mailing list or the PRs modifying the SLEP. The
possible paths of the status of SLEPs are as follows:

.. image:: pep-0001-process_flow.png
   :alt: SLEP process flow diagram

All SLEPs should be created with the ``Draft`` status.

Eventually, after discussion, there may be a consensus that the SLEP should be
accepted – see the next section for details. At this point the status becomes
``Accepted``.

Once a SLEP has been ``Accepted``, the reference implementation must be
completed. When the reference implementation is complete and incorporated into

> the main source code repository, the status will be changed to ``Final``.

Reviewer 2: Here, I think that we could add a section to give information about the scikit-learn version where the interface is considered stable.

Reviewer 2: xref: https://github.com/scikit-learn/enhancement_proposals/issues/56

Author: done, WDYT?

.. _slep_000:

==============================
SLEP000: SLEP and its workflow
==============================

:Author: Adrin Jalali
:Status: Draft
:Type: Process

> :Created: 2020-02-13

Reviewer 3: :fearful:

specification of the proposed solution, and a rationale for the feature.

We intend SLEPs to be the primary mechanisms for proposing major new features,
for collecting community input on an issue, and for documenting the design
decisions that have gone into scikit-learn. The SLEP author is responsible for
building consensus within the community and documenting dissenting opinions.

Because the SLEPs are maintained as text files in a versioned repository, their
revision history is the historical record of the feature proposal.

SLEP Audience
#############

The typical primary audience for SLEPs are the core developers of
``scikit-learn`` and technical committee, as well as contributors to the
project. However, these documents also serve the purpose of documenting the
changes and decisions to help users understand the changes and why they are
made. The SLEPs are available under `Scikit-learn enhancement proposals
<https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/>`_.

SLEP Types
##########

There are three kinds of SLEPs:

1. A Standards Track SLEP describes a new feature or implementation for
scikit-learn.

> 2. An Informational SLEP describes a scikit-learn design issue, or provides

Reviewer 3: Just for clarification, these are new, right? Does anyone want to give an example of these, and when they are useful? The notion is a bit vague for me tbh.

general guidelines or information to the scikit-learn community, but does not
propose a new feature. Informational SLEPs do not necessarily represent a
scikit-learn community consensus or recommendation, so users and implementers
are free to ignore Informational SLEPs or follow their advice.

3. A Process SLEP describes a process surrounding scikit-learn, or proposes a
change to (or an event in) a process. Process SLEPs are like Standards Track
SLEPs but apply to areas other than the scikit-learn library itself. They may
propose an implementation, but not to scikit-learn’s codebase; they require

> community consensus. Examples include procedures, guidelines, changes to the

Reviewer 3: should we explicitly include the governance document or is that encapsulated in the decision-making process?

deferring some features to later SLEPs), as this status can lead to version
compatibility challenges in the wider scikit-learn ecosystem.

A SLEP can also be assigned status ``Deferred``. The SLEP author or a core
developer can assign the SLEP this status when no progress is being made on the
SLEP.

A SLEP can also be ``Rejected``. Perhaps after all is said and done it was not
a good idea. It is still important to have a record of this fact. The
``Withdrawn`` status is similar; it means that the SLEP author themselves has
decided that the SLEP is actually a bad idea, or has accepted that a competing
proposal is a better alternative.

When a SLEP is ``Accepted``, ``Rejected``, or ``Withdrawn``, the SLEP should be
updated accordingly. In addition to updating the status field, at the very
least the ``Resolution`` header should be added with a link to the relevant
thread in the mailing list archives or where the discussion happened.

SLEPs can also be ``Superseded`` by a different SLEP, rendering the original
obsolete. The ``Replaced-By`` and ``Replaces`` headers should be added to the
original and new SLEPs respectively.

> ``Process`` SLEPs may also have a status of ``Active`` if they are never meant

Reviewer 6: My question is actually answered later:
> Process SLEPs may be updated over time to reflect changes to development practices and other details. The precise process followed in these cases will depend on the nature and purpose of the SLEP being updated.
